### DISCUSSION TITLE
Livewire Rules

### DISCUSSION THREAD

**Discussion by francoism90** [spatie/laravel-data#223] [Q&A]
I'm trying to use Livewire and laravel-data together, by using something like this:
```
```php
class PostEditController extends Component
{
    protected function rules(): array
    {
        return PostData::rules(); // this will return the full set - even lazy/exclude/include ones
    }
}
```
These are my data objects:
```
class PostData extends Data
{
    public function __construct(
        public string $name,
        public Lazy|UserData $user, // included in the rules object
    ) {
    }

    public static function fromModel(Post $model): self
    {
        return new self(
            $model->name,
            Lazy::create(fn() => UserData::from($model->user)),
        );
    }

    public static function rules(array $payload = []): array
    {
        return [
            'name' => [new Required(), new StringType(), new Max(40)],
            'user' => [new Sometimes(), new ArrayType()],
        ];
    }
}
```
```
class UserDataData extends Data
{
    public function __construct(
        public string $email,
    ) {
    }

    public static function fromModel(User $model): self
    {
        return new self(
            $model->email,
        );
    }

    public static function rules(array $payload = []): array
    {
        return [
            'email' => [new Required(), new EmailType(), new Max(40)],
        ];
    }
}
```
Would it be possible to build the rules according to what's actually in the payload? Any other workarounds?

Thanks!

**Comment by rubenvanassche** [comment]
Take a look at optional, Lazy is only being used for transforming a data object into a resource array. Not the other way around, so it shouldn't have affect on creating a data object or creating validation rules. When a property is optional, the sometimes rule gets added.

**Reply by francoism90** [reply]
@rubenvanassche Thanks for your reply. :)

The problem is that `PostData::rules()` returns all validation rules, including the ones from the lazy. The lazy isn't loaded, but still the rules from that one are included.

**Reply by francoism90** [reply]
@rubenvanassche This seems to be the cause: https://github.com/spatie/laravel-data/commit/12460ee690a40082e31b189e3b7caeeaa8b3dbf3

If I put it back to:
```
->reject(fn (DataProperty $property) => array_key_exists($property->name, $overWrittenRules) || ! $property->validate)
```
It seems to work fine. :)

**Reply by rubenvanassche** [reply]
Hi @francoism90,

This PR breaks a test, and normally the merge operation should replace generated rules by overwritten rules.

Please take a look at the Optional typehint, as said earlier, Lazy has nothing to do with validation rules and generation of validation rules will never depend on it.

For example this class:

```php
      $dataClass = new class('', SimpleData::fromString('Hello')) extends Data {
          public function __construct(
              public string $name,
              public Lazy|SimpleData $simple,
          )
          {
          }
      };
```

Will produce following rules:

```
[
  "name" => array:2 [
    0 => "string"
    1 => "required"
  ]
  "simple" => array:2 [
    0 => "required"
    1 => "array"
  ]
  "simple.string" => array:2 [
    0 => "string"
    1 => "required"
  ]
]
```

By swapping or adding Optional:

```php
        $dataClass = new class('', SimpleData::fromString('Hello')) extends Data {
            public function __construct(
                public string $name,
                public Lazy|Optional|SimpleData $simple,
            )
            {
            }
        };
```

The rules become:

```
[
  "name" => array:2 [
    0 => "string"
    1 => "required"
  ]
  "simple" => array:2 [
    0 => "sometimes"
    1 => "array"
  ]
  "simple.string" => array:2 [
    0 => "nullable"
    1 => "string"
  ]
]
```

**Reply by francoism90** [reply]
@rubenvanassche Thanks for your reply. :)

I've found a better alternative, adding `ExcludeWithout`. This will exclude the rules if the attribute does not take part in the payload. :)
I'll make a PR, as it seems this rule does not exist yet.

**Reply by francoism90** [reply]
@rubenvanassche Take this example:
```php
class UserData extends Data
{
    public function __construct(
        public string $name,
    )
    {
    }

    public static function rules(): array
    {
        return [
            'name' => [new Required(), new StringType(), new Max(255),
        ];
    }
}
```

```php
class PostData extends Data
{
    public function __construct(
        public string $name,
        public Lazy|Optional|UserData $user,
    )
    {
    }

    public static function rules(): array
    {
        return [
            'user' => [new RequiredWhen('user')],
        ];
    }
}
```

Payload:
```php
$post = PostData::validateAndCreate([
    'name' => 'Post 1',
]);
```

Result: 
```
user.name is required (even when no user has been given)
```

If I use `ExcludeWithout`, it works fine. :)

**Reply by rubenvanassche** [reply]
Yeah, this is a bigger problem than I thought.

The optional directive will convert the rules into something like this:

```php
[
  'user' => ['sometimes', ...],
  'user.name' => ['required', 'string', ...]
]
```

I would expect that when we provide a payload without user, that the `user.name` rules would not be executed since we clearly tell the validator that the user object is sometimes not being defined. That said, Laravel will execute the `user.name` rule, even if the user object is not given.

A solution would be to add a sometimes directive to this rule like this:

```php
[
  'user' => ['sometimes', ...],
  'user.name' => ['sometimes', 'required', 'string', ...]
]
```

This works, but actually breaks validation. If the user object also has a property age which always needs to be defined if the user object is given then using the approach above providing an empty user object will also be validated correctly.

I'll merge the PR you made probably later today but we've got a bigger problem here. And at the moment I have no idea how to fix it.

**Reply by francoism90** [reply]
@rubenvanassche I think the `exclude` rules help in this case. If I understand it correctly, you simple tell the validator ignore any sub-rules when no payload for it has been given.

I've tried `RequiredIf` and `sometimes`, however you'll still have the (sub)rules listed, they aren't 'scraped', if I'm not mistaken.

Maybe we should ask the Laravel devs?

**Reply by rubenvanassche** [reply]
I've created an issue on it: https://github.com/laravel/framework/issues/44586

We'll see what happens over there 😄
